File: Sum Array Puzzle/Sum_Array_Puzzle.py
class Solution:
    def sumArray(self, arr):
        # 1st solution
        total_sum = sum(arr)
        for i in range(len(arr)):
            arr[i] = total_sum - arr[i]
        
        # 2nd solution
        
        # n = len(arr)
        # prefix = [0] * n
        # suffix = [0] * n
        # total_sum = 0

        # for i in range(n):
        #     prefix[i] = total_sum
        #     total_sum += arr[i]

        # total_sum = 0
        # for i in range(n - 1, -1, -1):
        #     suffix[i] = total_sum
        #     total_sum += arr[i]

        # for i in range(n):
        #     arr[i] = prefix[i] + suffix[i]

File: Sum Array Puzzle/Sum_Array_Puzzle.cpp
class Solution {
  public:

    void sumArray(vector<int> &arr) {
        // 1st solution
        
        // int sum = 0;
        // for(auto i: arr) {
        //     sum += i;
        // }
        // for(auto &i: arr) {
        //     i = sum-i;
        // }
        // return;
        
        
        
        // 2nd solution
        
        int n = arr.size();
        vector<int> prefix(n,0), suffix(n,0);
        int sum =0;
        for(int i =0; i<n; i++) {
            prefix[i] = sum;
            sum += arr[i];
        }
        sum =0;
        for(int i =n-1; i>=0; i--) {
            suffix[i] =sum;
            sum += arr[i];
        }
        for(int i =0; i<n; i++) {
            arr[i] = prefix[i]+suffix[i];
        }
        return;
    }
};